pkg/sessions/dynamodb/dynamodb_store.go

```diff
@@ -0,0 +1,139 @@
+package dynamodb
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	dynamo "github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/sessions/persistence"
+)
+
+const SessionKeyName = "SessionKey"
+
+// GetPutDeleteWithContext is a subset of the dynamodbiface.DynamoDBAPI interface that defines only the operations
+// we need. This is useful for testing as it enables us to provide a custom client that doesn't have to implement the full interface.
+type GetPutDeleteWithContext interface {
+	GetItemWithContext(aws.Context, *dynamo.GetItemInput, ...request.Option) (*dynamo.GetItemOutput, error)
+	PutItemWithContext(aws.Context, *dynamo.PutItemInput, ...request.Option) (*dynamo.PutItemOutput, error)
+	DeleteItemWithContext(aws.Context, *dynamo.DeleteItemInput, ...request.Option) (*dynamo.DeleteItemOutput, error)
+}
+
+// SessionStore is an implementation of the persistence.Store
+// interface that stores sessions in dynamoDB
+type SessionStore struct {
+	dynamoService GetPutDeleteWithContext
+	tableName     string
+}
+
+type DynamoSessionItem struct {
+	SessionKey string
+	Value      []byte
+	Expiry     string
+}
+
+var now = time.Now
+
+func NewDynamoDBSessionStore(opts *options.SessionOptions, cookieOpts *options.Cookie) (sessions.SessionStore, error) {
+	sess, err := session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to dynamodb: %v", err)
+	}
+
+	return persistence.NewManager(newSessionStore(dynamo.New(sess), opts.DynamoDB.TableName), cookieOpts), nil
+}
+
+func newSessionStore(client GetPutDeleteWithContext, tableName string) persistence.Store {
+	ss := &SessionStore{
+		dynamoService: client,
+		tableName:     tableName,
+	}
+
+	return ss
+}
+
+func (store *SessionStore) Save(ctx context.Context, key string, bytes []byte, duration time.Duration) error {
+	av, err := dynamodbattribute.MarshalMap(DynamoSessionItem{
+		SessionKey: key,
+		Value:      bytes,
+		Expiry:     strconv.FormatInt(time.Now().Add(duration).Unix(), 10),
```

Comment:
This might be a dumb question because I'm less familiar with DynamoDB nitty gritty:

We seem to always want to evaluate time as an integer, but we cast & uncast from a String to fit the string type for `Expiry` in `DynamoSessionItem`.  Can `Expiry` just be a numeric type from the get-go?

Reply:
And how do we set the actual TTL on the DynamoDB items so the slow async AWS side cleanup happens behind the scenes so our session store doesn't balloon in size over time?

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/time-to-live-ttl-before-you-start.html#time-to-live-ttl-before-you-start-formatting

Reply:
Seems like once we align this we will need to add documentation for users to configure the TTL field name we choose on their own for their DynamoDB table they provide us?